File: clients/elixir/lib/transport.ex
defmodule Corndogs.Transport do
  @moduledoc """
  Corndogs' own, ready-to-use client transport: CSIL-RPC over a persistent TCP
  connection. You do not need to write a carrier yourself — point this at a
  corndogs server's TCP address and hand it to the generated client:

      transport = Corndogs.Transport.connect("localhost:5080")
      Corndogs.Transport.start_heartbeat(transport)   # keep the connection alive (background)
      client = Csilgen.Generated.CorndogsClient.new(transport)

      resp =
        Csilgen.Generated.CorndogsClient.submit_task(client, %Csilgen.Generated.SubmitTaskRequest{
          queue: "emails",
          current_state: "submitted",
          auto_target_state: "sending",
          timeout: -1,
          payload: "...",
          priority: 0
        })

  The wire is CSIL-RPC framed with a 4-byte big-endian length prefix over
  `:gen_tcp` (no `:inets`/`:ssl` dependency, so it runs as-is under `mix` or a
  release with no `extra_applications`). This module owns only the envelope,
  framing, and connection lifecycle; the generated codec owns
  (de)serialization of typed request/response structs.

  The connection is a `GenServer` that owns the socket, so calls are
  serialized and automatically re-dial on failure. It implements
  `Csilgen.Generated.Transport`, the behaviour the generated client drives.
  """

  use GenServer

  @behaviour Csilgen.Generated.Transport

  alias Csilgen.Generated.Cbor
  alias Csilgen.Generated.ServiceError, as: GeneratedServiceError

  # RFC 8949 §3.4.5.1 — embedded encoded CBOR data item.
  @tag_encoded_cbor 24
  @max_frame 256 * 1024 * 1024
  @default_connect_timeout 5_000
  @default_heartbeat_interval 15_000

  # Control-plane heartbeat: never collides with an application op.
  @control_service "CorndogsService"
  @op_ping "$ping"

  defstruct [:pid]
  @type t :: %__MODULE__{pid: pid()}

  defmodule TransportError do
    @moduledoc "A transport-level failure: connection dropped, dial failed, malformed frame."
    defexception [:message]
  end

  defmodule ServiceError do
    @moduledoc "A structured application error the service returned."
    defexception [:code, :message]

    @impl true
    def message(%__MODULE__{code: code, message: msg}),
      do: "corndogs: service error #{code}: #{msg}"
  end

  # --- construction ------------------------------------------------------

  @doc """
  Connects a new transport to `addr` (`"host:port"`). The socket is dialed
  lazily on the first call, and re-dialed automatically after a failure.
  """
  @spec connect(String.t(), keyword()) :: t()
  def connect(addr, opts \\ []) do
    {:ok, pid} = GenServer.start_link(__MODULE__, {addr, opts})
    %__MODULE__{pid: pid}
  end

  @doc "Closes the connection (and stops the socket)."
  @spec close(t()) :: :ok
  def close(%__MODULE__{pid: pid}) do
    if Process.alive?(pid), do: GenServer.stop(pid)
    :ok
  end

  # --- Csilgen.Generated.Transport behaviour ------------------------------

  @doc """
  Implements `Csilgen.Generated.Transport.call/4`: sends one request and
  waits for its correlated response. Raises `Corndogs.Transport.ServiceError`
  for a typed application error, or `Corndogs.Transport.TransportError` for a
  connection/transport failure.
  """
  @impl Csilgen.Generated.Transport
  @spec call(t(), String.t(), String.t(), binary()) :: binary()
  def call(%__MODULE__{pid: pid}, service, op, req) do
    case GenServer.call(pid, {:rpc, service, op, req}, :infinity) do
      {:ok, payload} ->
        payload

      {:error, %GeneratedServiceError{code: code, message: message}} ->
        raise ServiceError, code: code, message: message

      {:error, reason} ->
        raise TransportError, message: format_reason(reason)
    end
  end

  # --- heartbeat -----------------------------------------------------------

  @doc "Sends a single control-plane heartbeat; raises if the server is unreachable."
  @spec ping(t()) :: :ok
  def ping(%__MODULE__{} = t) do
    call(t, @control_service, @op_ping, <<>>)
    :ok
  end

  @doc """
  SYNCHRONOUS heartbeat: blocks the calling process, pinging every `interval`
  milliseconds, until either a ping fails (returns `{:error, exception}`) or
  the process receives `:corndogs_stop_heartbeat` (returns `:ok`). Run it in a
  process you own (e.g. `spawn/1` or `Task.start/1`) and `send/2` it the stop
  message to end it — or use `start_heartbeat/2` for the background form.
  """
  @spec run_heartbeat(t(), non_neg_integer()) :: :ok | {:error, Exception.t()}
  def run_heartbeat(%__MODULE__{} = t, interval \\ @default_heartbeat_interval) do
    receive do
      :corndogs_stop_heartbeat -> :ok
    after
      interval ->
        try do
          ping(t)
          run_heartbeat(t, interval)
        rescue
          e -> {:error, e}
        end
    end
  end

  @doc """
  ASYNCHRONOUS heartbeat: runs `run_heartbeat/2` on a background process and
  returns a zero-arity stop function. Calling it stops the heartbeat; a failed
  ping also ends it on its own (the connection re-dials on the next real call
  regardless).
  """
  @spec start_heartbeat(t(), non_neg_integer()) :: (-> :ok)
  def start_heartbeat(%__MODULE__{} = t, interval \\ @default_heartbeat_interval) do
    pid = spawn(fn -> run_heartbeat(t, interval) end)
    fn -> send(pid, :corndogs_stop_heartbeat) end
  end

  # --- GenServer: owns the socket, serializes calls, re-dials on failure ---

  @impl GenServer
  def init({addr, opts}) do
    {host, port} = split_addr(addr)
    timeout = Keyword.get(opts, :connect_timeout, @default_connect_timeout)
    {:ok, %{host: host, port: port, timeout: timeout, socket: nil, next_id: 0}}
  end

  @impl GenServer
  def handle_call({:rpc, service, op, req}, _from, state) do
    id = state.next_id + 1
    state = %{state | next_id: id}

    case ensure_socket(state) do
      {:ok, socket, state} ->
        case send_and_recv(socket, encode_envelope(service, op, id, req)) do
          {:ok, frame} ->
            {:reply, parse_response(frame), state}

          {:error, reason} ->
            close_socket(socket)
            {:reply, {:error, reason}, %{state | socket: nil}}
        end

      {:error, reason} ->
        {:reply, {:error, reason}, state}
    end
  end

  @impl GenServer
  def terminate(_reason, state) do
    close_socket(state.socket)
    :ok
  end

  defp ensure_socket(%{socket: nil, host: host, port: port, timeout: timeout} = state) do
    case :gen_tcp.connect(host, port, [:binary, active: false, packet: :raw], timeout) do
      {:ok, socket} ->
        :inet.setopts(socket, nodelay: true, keepalive: true)
        {:ok, socket, %{state | socket: socket}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp ensure_socket(%{socket: socket} = state), do: {:ok, socket, state}

  defp close_socket(nil), do: :ok
  defp close_socket(socket), do: :gen_tcp.close(socket)

  defp send_and_recv(socket, env) do
    with :ok <- write_frame(socket, env) do
      read_frame(socket)
    end
  end

  defp write_frame(_socket, env) when byte_size(env) > @max_frame,
    do: {:error, {:frame_too_large, byte_size(env)}}

  defp write_frame(socket, env) do
    :gen_tcp.send(socket, <<byte_size(env)::big-unsigned-integer-size(32), env::binary>>)
  end

  defp read_frame(socket) do
    with {:ok, len_bin} <- :gen_tcp.recv(socket, 4) do
      <<len::big-unsigned-integer-size(32)>> = len_bin

      cond do
        len > @max_frame -> {:error, {:frame_too_large, len}}
        len == 0 -> {:ok, <<>>}
        true -> :gen_tcp.recv(socket, len)
      end
    end
  end

  # --- envelope -------------------------------------------------------------

  defp encode_envelope(service, op, id, req) do
    Cbor.encode(
      {:map,
       [
         {{:text, "v"}, {:int, 1}},
         {{:text, "service"}, {:text, service}},
         {{:text, "op"}, {:text, op}},
         {{:text, "id"}, {:int, id}},
         {{:text, "payload"}, {:tag, @tag_encoded_cbor, {:bytes, req}}}
       ]}
    )
  end

  # Decodes a CsilRpcResponse frame into {:ok, inner_payload} | {:error, reason}. A
  # non-zero transport status or a "ServiceError" variant becomes an {:error, _}.
  defp parse_response(frame) do
    case Cbor.decode(frame) do
      {:map, _} = env ->
        status = map_get_int(env, "status", 0)

        if status != 0 do
          {:error, {:transport_status, status, map_get_text(env, "error", "")}}
        else
          decode_payload(env)
        end

      _ ->
        {:error, :malformed_response}
    end
  rescue
    e -> {:error, {:malformed_response, e}}
  end

  defp decode_payload(env) do
    case map_get(env, "payload") do
      nil ->
        {:ok, <<>>}

      {:tag, _num, {:bytes, inner}} ->
        wrap_variant(env, inner)

      {:bytes, inner} ->
        wrap_variant(env, inner)

      _ ->
        {:error, :malformed_payload}
    end
  end

  defp wrap_variant(env, inner) do
    if map_get_text(env, "variant", nil) == "ServiceError" do
      {:error, GeneratedServiceError.from_cbor(inner)}
    else
      {:ok, inner}
    end
  end

  defp map_get({:map, kvs}, key) do
    Enum.find_value(kvs, fn
      {{:text, ^key}, v} -> {:ok, v}
      _ -> nil
    end)
    |> case do
      {:ok, v} -> v
      nil -> nil
    end
  end

  defp map_get_int(env, key, default) do
    case map_get(env, key) do
      {:int, n} -> n
      _ -> default
    end
  end

  defp map_get_text(env, key, default) do
    case map_get(env, key) do
      {:text, s} -> s
      _ -> default
    end
  end

  # --- address parsing --------------------------------------------------

  defp split_addr(addr) do
    parts = String.split(addr, ":")

    case Enum.split(parts, -1) do
      {host_parts, [port_str]} when host_parts != [] ->
        {String.to_charlist(Enum.join(host_parts, ":")), String.to_integer(port_str)}

      _ ->
        raise ArgumentError, "corndogs: address must be host:port, got #{inspect(addr)}"
    end
  end

  defp format_reason({:transport_status, status, msg}), do: "transport status #{status}: #{msg}"
  defp format_reason({:frame_too_large, n}), do: "frame too large (#{n})"
  defp format_reason(:malformed_response), do: "malformed response envelope"
  defp format_reason(:malformed_payload), do: "response payload is not a tag-24 byte string"

  defp format_reason({:malformed_response, e}),
    do: "malformed response envelope: #{Exception.message(e)}"

  defp format_reason(:closed), do: "connection closed"
  defp format_reason(reason) when is_atom(reason), do: :inet.format_error(reason) |> to_string()
  defp format_reason(reason), do: inspect(reason)

  @doc "Convenience: a ready transport for `addr` (`\"host:port\"`)."
  @spec new(String.t(), keyword()) :: t()
  def new(addr, opts \\ []), do: connect(addr, opts)
end

File: clients/dart/lib/transport.dart
// Corndogs client transport: CSIL-RPC over a persistent TCP connection.
//
// This is Corndogs' own, ready-to-use carrier — you do not need to write one or
// read csilgen docs. Point it at a corndogs server's TCP address and pass it to
// the generated async client:
//
//   import 'package:corndogs/corndogs.dart';
//   import 'package:corndogs/transport.dart';
//
//   final tr = await TcpTransport.connect('localhost:5080');
//   tr.startHeartbeat(const Duration(seconds: 15)); // keep the connection alive
//   final client = CorndogsAsyncClient(tr);
//   final resp = await client.submitTask(SubmitTaskRequest(
//     queue: 'q', currentState: 'submitted', autoTargetState: 'sending',
//     timeout: -1, payload: Uint8List(0), priority: 0,
//   ));
//
// The wire is CSIL-RPC framed with the canonical 4-byte big-endian length prefix
// over TCP. This carrier owns only the envelope + framing + connection; the
// generated client owns (de)serialization. Dependency-free (dart:io + dart:async
// only; reuses the generated CBOR codec for every value it can).
//
// One connection MULTIPLEXES many concurrent calls: each request carries a
// correlation `id`, a background frame reader matches each response back to its
// waiting caller, so concurrent submitTask/getNextTask calls do not head-of-line
// block each other.
//
// Dart has no synchronous socket API, so only the generated *async* client
// (`CorndogsAsyncClient`, via `AsyncCsilTransport`) is served here. The generated
// synchronous `CorndogsClient` (`CsilTransport`) cannot be backed by real network
// I/O in Dart and is intentionally not implemented by this carrier.

import 'dart:async';
import 'dart:io';
import 'dart:typed_data';

import 'csil_cbor.gen.dart';
import 'client.async.gen.dart';
import 'types.gen.dart';

const int _tagEncodedCbor = 24; // RFC 8949 §3.4.5.1 — embedded encoded CBOR data item
const int _maxFrame = 1025 << 20; // payload maximum plus envelope allowance
const Duration _defaultConnectTimeout = Duration(seconds: 5);
const Duration _defaultHeartbeatInterval = Duration(seconds: 15);
const String _controlService = 'CorndogsService';
const String _opPing = r'$ping'; // control-plane heartbeat op (never collides with app ops)

/// A transport-level failure: connection lost, malformed frame, a non-zero
/// transport status, or an invalid address/timeout. Distinct from
/// [CorndogsServiceException], which carries a structured application error.
class CorndogsTransportException implements Exception {
  final String message;
  const CorndogsTransportException(this.message);

  @override
  String toString() => 'CorndogsTransportException: $message';
}

/// A structured application error the server returned (CSIL-RPC's
/// `ServiceError` variant), decoded via the generated [ServiceError] type.
class CorndogsServiceException implements Exception {
  final ServiceError error;
  const CorndogsServiceException(this.error);

  int get code => error.code;
  String get message => error.message;

  @override
  String toString() => 'CorndogsServiceException($code): $message';
}

/// CSIL-RPC transport over a persistent, multiplexed TCP connection. Implements
/// the generated [AsyncCsilTransport] protocol (`call`), so it plugs straight
/// into [CorndogsAsyncClient]. The connection is dialed lazily (on first call,
/// or eagerly via [TcpTransport.connect]) and re-dialed on failure.
class TcpTransport implements AsyncCsilTransport {
  /// A lazily-connecting transport for `host:port`. The first call to [call]
  /// (or [ping]) dials the connection. Use [TcpTransport.connect] instead if
  /// you want to await the connection up front.
  TcpTransport(String addr, {Duration connectTimeout = _defaultConnectTimeout})
      : _host = _splitAddr(addr).$1,
        _port = _splitAddr(addr).$2,
        _connectTimeout = connectTimeout;

  final String _host;
  final int _port;
  final Duration _connectTimeout;

  Socket? _socket;
  StreamSubscription<Uint8List>? _sub;
  Future<void>? _connecting;
  bool _closed = false;

  final List<int> _recvBuf = <int>[];
  final Map<int, Completer<Object?>> _pending = <int, Completer<Object?>>{};
  int _nextId = 0;

  Timer? _hbTimer;

  /// An already-connected [TcpTransport] for `addr` (`host:port`).
  static Future<TcpTransport> connect(
    String addr, {
    Duration connectTimeout = _defaultConnectTimeout,
  }) async {
    final t = TcpTransport(addr, connectTimeout: connectTimeout);
    await t._ensureConnected();
    return t;
  }

  Future<void> _ensureConnected() {
    if (_closed) {
      throw const CorndogsTransportException('transport closed');
    }
    if (_socket != null) return Future.value();
    final inFlight = _connecting;
    if (inFlight != null) return inFlight;

    final completer = Completer<void>();
    _connecting = completer.future;
    () async {
      try {
        final socket = await Socket.connect(_host, _port, timeout: _connectTimeout);
        socket.setOption(SocketOption.tcpNoDelay, true);
        _recvBuf.clear();
        _socket = socket;
        _sub = socket.listen(
          _onData,
          onDone: () => _teardown(const CorndogsTransportException('connection closed')),
          onError: (Object e) => _teardown(CorndogsTransportException('socket error: $e')),
          cancelOnError: true,
        );
        completer.complete();
      } catch (e) {
        completer.completeError(CorndogsTransportException('connect failed: $e'));
      } finally {
        _connecting = null;
      }
    }();
    return completer.future;
  }

  void _onData(List<int> chunk) {
    _recvBuf.addAll(chunk);
    while (true) {
      if (_recvBuf.length < 4) return;
      final len = (_recvBuf[0] << 24) | (_recvBuf[1] << 16) | (_recvBuf[2] << 8) | _recvBuf[3];
      if (len > _maxFrame) {
        _teardown(CorndogsTransportException('frame too large ($len)'));
        return;
      }
      if (_recvBuf.length < 4 + len) return;
      final frame = Uint8List.fromList(_recvBuf.sublist(4, 4 + len));
      _recvBuf.removeRange(0, 4 + len);
      _handleFrame(frame);
    }
  }

  void _handleFrame(Uint8List frame) {
    Object? val;
    try {
      val = CsilCbor.decode(frame);
    } catch (_) {
      return; // malformed frame: drop it, the caller's request will eventually time out
    }
    if (val is! Map) return;
    final idVal = val['id'];
    final id = idVal is int ? idVal : 0;
    final completer = _pending.remove(id);
    if (completer != null && !completer.isCompleted) {
      completer.complete(val);
    }
  }

  /// Closes the current connection (if it is still the live one) and fails every
  /// pending call, so the next call re-dials.
  void _teardown(CorndogsTransportException cause) {
    final socket = _socket;
    if (socket == null) return;
    _socket = null;
    _sub?.cancel();
    _sub = null;
    _recvBuf.clear();
    socket.destroy();
    final pending = Map<int, Completer<Object?>>.from(_pending);
    _pending.clear();
    for (final c in pending.values) {
      if (!c.isCompleted) c.completeError(cause);
    }
  }

  @override
  Future<Uint8List> call(String service, String op, List<int> request) async {
    await _ensureConnected();
    final socket = _socket;
    if (socket == null) {
      throw const CorndogsTransportException('connection lost');
    }

    final id = ++_nextId;
    final completer = Completer<Object?>();
    _pending[id] = completer;

    final envelope = _buildRequestEnvelope(service, op, id, request);
    try {
      await _writeFrame(socket, _withLengthPrefix(envelope));
    } catch (e) {
      _pending.remove(id);
      final err = CorndogsTransportException('write failed: $e');
      _teardown(err);
      throw err;
    }

    final val = await completer.future;
    return _extractPayload(val);
  }

  // Serializes frame writes on the shared connection: concurrent calls each add
  // + flush the same Socket, and Dart's IOSink throws if a second flush starts
  // before the first completes, so writes are chained through this queue. Reads
  // (and therefore concurrent calls in flight) are unaffected — only the write
  // side is serialized.
  Future<void> _writeQueue = Future<void>.value();

  Future<void> _writeFrame(Socket socket, Uint8List frame) {
    final next = _writeQueue.then((_) async {
      socket.add(frame);
      await socket.flush();
    });
    // Swallow the error here so a failed write doesn't leave `_writeQueue`
    // permanently rejected for the next call; the caller above still sees it.
    _writeQueue = next.catchError((Object _) {});
    return next;
  }

  Uint8List _extractPayload(Object? val) {
    if (val is! Map) {
      throw const CorndogsTransportException('malformed response envelope');
    }
    final status = val['status'];
    if (status is int && status != 0) {
      final err = val['error'];
      throw CorndogsTransportException('transport status $status: ${err ?? ''}');
    }
    final payload = val['payload'];
    if (payload == null) {
      return Uint8List(0); // control replies (e.g. $pong) carry no payload
    }
    final bytes = payload is Uint8List ? payload : Uint8List.fromList((payload as List).cast<int>());
    if (val['variant'] == 'ServiceError') {
      throw CorndogsServiceException(ServiceError.fromCborValue(CsilCbor.decode(bytes)));
    }
    return bytes;
  }

  // --- heartbeat -------------------------------------------------------------

  /// Sends a single control-plane heartbeat; throws if the server is
  /// unreachable. Cheap; keeps an idle connection alive and detects a dead
  /// server.
  Future<void> ping() async {
    await call(_controlService, _opPing, const <int>[]);
  }

  /// The SYNCHRONOUS-style heartbeat start: an awaitable loop that pings every
  /// [interval] until [close] is called (returns normally) or a ping fails
  /// (the error propagates out of the awaited future). Await it in a context
  /// you own, or use [startHeartbeat] for the background form.
  Future<void> runHeartbeat([Duration interval = _defaultHeartbeatInterval]) async {
    while (!_closed) {
      await Future<void>.delayed(interval);
      if (_closed) return;
      await ping();
    }
  }

  /// The ASYNCHRONOUS heartbeat start: runs the heartbeat on a background
  /// [Timer] and returns a stop function. Calling the returned function (or
  /// [close]) ends it. Only one background heartbeat runs at a time; a repeat
  /// call returns a stop function for the existing one.
  void Function() startHeartbeat([Duration interval = _defaultHeartbeatInterval]) {
    final existing = _hbTimer;
    if (existing != null) {
      return () => _stopHeartbeatTimer(existing);
    }
    final timer = Timer.periodic(interval, (_) {
      unawaited(ping().catchError((Object _) {})); // a dead server just means the next tick retries
    });
    _hbTimer = timer;
    return () => _stopHeartbeatTimer(timer);
  }

  void _stopHeartbeatTimer(Timer timer) {
    timer.cancel();
    if (identical(_hbTimer, timer)) _hbTimer = null;
  }

  /// Shuts the transport down: stops any heartbeat, fails pending calls, and
  /// closes the connection.
  Future<void> close() async {
    _closed = true;
    _hbTimer?.cancel();
    _hbTimer = null;
    _teardown(const CorndogsTransportException('transport closed'));
  }
}

/// Convenience: an already-connected [TcpTransport] for `addr` (`host:port`).
Future<TcpTransport> connect(String addr) => TcpTransport.connect(addr);

(String, int) _splitAddr(String addr) {
  final idx = addr.lastIndexOf(':');
  if (idx <= 0 || idx == addr.length - 1) {
    throw ArgumentError('corndogs: address must be host:port, got $addr');
  }
  final host = addr.substring(0, idx);
  final port = int.tryParse(addr.substring(idx + 1));
  if (port == null) {
    throw ArgumentError('corndogs: address must be host:port, got $addr');
  }
  return (host, port);
}

// --- envelope + framing helpers ---------------------------------------------
//
// The generated CsilCbor codec (csil_cbor.gen.dart) encodes any single scalar,
// list, or map value, so it is reused for every value below. It does not expose
// a builder for a CBOR tag (major type 6) — needed here to wrap the request
// payload in tag 24 ("embedded CBOR data item") — so the handful of bytes that
// make up that wrapper, and the fixed 5-key envelope map header, are written
// directly, mirroring CsilCbor's own canonical head-byte encoding.

Uint8List _buildRequestEnvelope(String service, String op, int id, List<int> payload) {
  final out = BytesBuilder();
  _writeHead(out, 5, 5); // map, 5 key/value pairs
  out.add(CsilCbor.encodeValue('v'));
  out.add(CsilCbor.encodeValue(1));
  out.add(CsilCbor.encodeValue('service'));
  out.add(CsilCbor.encodeValue(service));
  out.add(CsilCbor.encodeValue('op'));
  out.add(CsilCbor.encodeValue(op));
  out.add(CsilCbor.encodeValue('id'));
  out.add(CsilCbor.encodeValue(id));
  out.add(CsilCbor.encodeValue('payload'));
  _writeHead(out, 6, _tagEncodedCbor); // tag 24
  out.add(CsilCbor.encodeValue(Uint8List.fromList(payload))); // the tagged byte string
  return out.toBytes();
}

void _writeHead(BytesBuilder out, int major, int n) {
  final mt = major << 5;
  if (n < 24) {
    out.addByte(mt | n);
  } else if (n < 0x100) {
    out.addByte(mt | 24);
    out.addByte(n);
  } else if (n < 0x10000) {
    out.addByte(mt | 25);
    out.addByte((n >> 8) & 0xff);
    out.addByte(n & 0xff);
  } else {
    out.addByte(mt | 26);
    for (var i = 3; i >= 0; i--) {
      out.addByte((n >> (i * 8)) & 0xff);
    }
  }
}

Uint8List _withLengthPrefix(Uint8List body) {
  if (body.length > _maxFrame) {
    throw CorndogsTransportException('frame too large (${body.length} > $_maxFrame)');
  }
  final out = Uint8List(4 + body.length);
  ByteData.sublistView(out).setUint32(0, body.length, Endian.big);
  out.setRange(4, 4 + body.length, body);
  return out;
}
